### DISCUSSION TITLE
Should delta t be fixed?

### DISCUSSION THREAD

**Discussion by eeeewwqq** [JSBSim-Team/jsbsim#915] [Q&A]
Hi! I'm developing a game. Recently I made one major change. I set delta t on each step of simulation to value = time between to frames. Sometimes, when fps drops(delta t grows), the plane start jump on direct surface. Can floating delta t cause this? Is it valid to set delta t this way?

**Comment by gallonmate** [comment]
Yes the wheels/gear/spring sim can become unstable at lower simulation frequencies. JSBSim is recommended to run at 120hz. And it's also recommended to run a constant delta, not to use use a variable rate https://github.com/JSBSim-Team/jsbsim/discussions/449#discussioncomment-912346.

I wrote a little ugly code to run the Unreal Engine Plugin at a constant 120hz, independent of game rate. It works great at any game rate, even wildly fluctuating rates. The code below is run on each game tick/frame. Feel free to use it:

https://github.com/JSBSim-Team/jsbsim/blob/868aed43cb9df1f65fe29c426427c83fa4f86549/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L257

```c++
      //calculate sim rate to be 120hz independent of game tick rate(Pseudo fixed rate, dev/user needs to set game to a fixed rate too)
      simDtime = 120.f/(1.f / DeltaTime); // DeltaTime is Unreal Engine's each current frame delta time.
      remainder = remainder + fmodf(simDtime,1);
      simloops = truncf(simDtime) + truncf(remainder);
      remainder = fmodf(remainder,1);

      //jsbsim recommends to step sim at 120hz, eg 1/120 = 0.0083..
      Exec->Setdt(0.008333333333333333);

      //Send Commands and State to JSBSim
      CopyToJSBSim();
			
      //step sim x times per game tick
      for (int i = 0; i < simloops; i++)
      {
        Exec->Run();
      }

      //Get the results from JSBSim
      CopyFromJSBSim();
```

**Reply by eeeewwqq** [reply]
I'm writing ugly code for unity now:)

I don't know why, but in my case all this doesn't work. I think it because of the game view. It's 2.5d game, plane is 3d, background 2d, and this small inaccuracies are perfectly visible on 2d background.

I tried a lot of methods and come to this: I still using floating dT, but I limited max dT to 1/120. I run simulation for each frame required number of times with max dT + one run with small dT to get perfect accuracy. It's like yours algorithm, I just set reminder as dT and call extra run.

Thanks for reply.

**Reply by bcoconni** [reply]
@gallonmate, just a GitHub tip: if you want to include a code extract of several lines just append the path to the file with a dash and the line number of the last line of code you want to mention.

For example to display the code from line 257 to 267, just append `#L257-L267` to the end of the address such as:
`https://github.com/JSBSim-Team/jsbsim/blob/868aed43cb9df1f65fe29c426427c83fa4f86549/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L257-L267`

which displays
https://github.com/JSBSim-Team/jsbsim/blob/868aed43cb9df1f65fe29c426427c83fa4f86549/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L257-L267

**Reply by gallonmate** [reply]
Thank, always appreciate tips/help from the pros! Lol this time was on purpose though, I wanted to edit the code a little for easier reading but also give the file link. I agree it looks strange the way I posted both. :sweat_smile:

**Comment by seanmcleod** [comment]
Or what I typically do is browse to the relevant section of the code on Github, and then left-click on the starting line and then shift-left-click on the last line which then highlights the section of code, and then select the "Copy permalink" option from the context menu.

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/bfece95e-a75b-4225-bc69-7da456564307)

**Reply by bcoconni** [reply]
I did not know the "shift-left-click" trick.
Thanks for the tip ! :+1:
